playtime/init.lua

```diff
@@ -0,0 +1,88 @@
+playtime = {}
+
+local os = os
+local math = math
+local string = string
+
+local current = {}
+
+local key = "playtime:time"
+
+local S = minetest.get_translator("playtime")
+local C = minetest.colorize
+
+--get playtime
+function playtime.get_current_playtime(name)
+	if current[name] then
+		return os.time() - current[name]
+	else
+		return 0
+	end
+end
+
+--get total playtime
+function playtime.get_total_playtime(name)
+	local player = minetest.get_player_by_name(name)
+	if player then
+		return player:get_meta():get_int(key) + playtime.get_current_playtime(name)
+	end
+end
+
+--reset playtime
+function playtime.remove_playtime(name)
+	local player = minetest.get_player_by_name(name)
+	if player then
+		player:get_meta():set_int(key, 0)
+	end
+end
+
+if minetest.is_singleplayer() then
+	minetest.register_on_shutdown(function()
+		local player = minetest.get_connected_players()[1]
+		if player then
+			local name = player:get_player_name()
+			local meta = player:get_meta()
+			meta:set_int(key, meta:get_int(key) + playtime.get_current_playtime(name))
+			current[name] = nil
+		end
+	end)
+else
+	minetest.register_on_leaveplayer(function(player)
+		local name = player:get_player_name()
+		local meta = player:get_meta()
+		meta:set_int(key, meta:get_int(key) + playtime.get_current_playtime(name))
```

Comment:
I don't like this approach to persistence; session playtime is lost if the server crashes, for instance. I believe both values should be stored in meta. The session playtime would then be added to the total playtime and reset in `on_joinplayer`.

For higher efficiency of this approach, consider storing only the gametime of the last login to compute the session playtime as current gametime - last login gametime.

Reply:
Will look into it